Replace docker-compose

Saiv46 opened the pull request:
docker-compose file of this PR was meant to have little improvements over original, but now it's based on production config of [mastodon.ml](https://mastodon.ml)

**BLOCKED BY #23887 (this PR uses postgres-15, which is not being automatically tested yet)**

Major differences:
- Uses volumes for configurations and storage (`configs`/`secrets` only work in Swarm mode)
- Added network aliases (for use instead of IP)
- Removed Tor federation (too much trouble)
- Scaled by default, see [Scaling up your server](https://docs.joinmastodon.org/admin/scaling/)
- Optional metrics collection with Prometheus

Minor differences:
- Updated container images
- Using global volume for certificates (for other stacks to use)
- Self-hosted mail server ([Maddy](https://maddy.email/))
- Some containers are configured by environment variables only
- `stack.env` instead of `.env.production` (compatibility with Portainer interface)

Saiv46 committed:
Improve docker-compose

Saiv46 committed:
Returned Tor

danger89 commented:
This looks good!

BtbN commented:
Some things I've noticed:

You rename the environment file from .env.production to stack.env, without also updating the documentation to reflect that new path. Is that intended? Why the change in the first place?

The nginx container mounts ./nginx, but no such directory is to be seen. Same for the letsencrypt one.
Also, if you use nginx already, it could potentially be used to serve the static files. For which it'd need a second Dockerfile to copy them into there, since mounting them as a volume could lead to a mismatch between the running containers on updates.

I'm not sure I'm a fan of the compose file claiming ports 80 and 443 for itself, since at least in my case the whole setup is running behind a pre-existing reverse proxy, so those ports are already occupied.

danger89 commented:
> I'm not sure I'm a fan of the compose file claiming ports 80 and 443 for itself, since at least in my case the whole setup is running behind a pre-existing reverse proxy, so those ports are already occupied.

This I fully agree. The nginx service should be optional in the docker compose config.

BtbN commented:
Also, you are running two replicas of redis and sidekiq. Does that work just like that? I'd imagine that you at least need to somehow configure replication between the two instances of each service? Or does that automagically happen somehow?

danger89 commented:
> 

I think so: https://www.reddit.com/r/rails/comments/9db946/can_we_scale_sidekiq_with_containers/

BtbN commented:
So sidekiq might actually be fine, yeah. But everything I find online in regards to redis looks a lot more involved.
Like, I keep finding setups similar to what's outlined here: https://gist.github.com/nimasrn/eeb8d31855f55af570911610970bd192

Not quite sure how load balancing between them works either. It seems to be more for replication in case of machine failure?

nberlee commented:
I don't think sidekiq is fine as both replicas have the scheduler role, and there only should be one (the rest of the roles are fine to scale up)

Also, there really needs to be a note on the volume storage plugin. As by default docker stack (swarm) uses local, which is a problem when you share volumes on containers that are on different nodes.

Znuff commented:
> Use volumes instead of directories

Can you explain why volumes would benefit Mastodon in particular.

I understand the benefit when running it with a container orchestration, but most people don't have such a setup and it sounds like they would just end up in a world of hurt when they'll eventually want to move or scale their setups.

> I'm not sure I'm a fan of the compose file claiming ports 80 and 443 for itself, since at least in my case the whole setup is running behind a pre-existing reverse proxy, so those ports are already occupied.

I second this.

Saiv46 committed:
Update defaults of docker-compose

- Use postresql v15 (#21059)
- Make nginx/letsencrypt optional
- Run only one redis/sidekiq instance by default

Saiv46 committed:
Prettier error, duh

Saiv46 committed:
Merge branch 'main' into patch-1

Saiv46 committed:
Revamped docker-compose.yml

Saiv46 commented on docker-compose.yml:
```diff
@@ -1,91 +1,115 @@
   db:
     restart: always
-    image: postgres:14-alpine
+    image: postgres:15-alpine
```
Most monolithic instances still run on version 10 or older. After upgrading to 15, no problem so far. I hope CI will update soon
